File: Java/520. Detect Capital.java
/**
 * https://leetcode.com/problems/detect-capital/
 */

class Solution {
    // O(n) time complexity
    // O(1) time complexity
    public boolean detectCapitalUse(String word) {
        return isAllUpper(word) || isAllLower(word) || isFirstUpper(word);
    }

    private boolean isAllUpper(String word) {
        for (char c : word.toCharArray()) {
            if (!Character.isUpperCase(c))
                return false;
        }
        return true;
    }

    private boolean isAllLower(String word) {
        for (char c : word.toCharArray()) {
            if (!Character.isLowerCase(c))
                return false;
        }
        return true;
    }

    private boolean isFirstUpper(String word) {
        if (Character.isUpperCase(word.charAt(0))) {
            for (int i = 1; i < word.length(); i++) {
                if (Character.isUpperCase(word.charAt(i))) {
                    return false;
                }
            }
            return true;
        } else {
            return false;
        }
    }
}

File: Kotlin/520. Detect Capital.kt
/**
 *  https://leetcode.com/problems/detect-capital/
 */

class Solution {
    // O(n) time complexity
    // O(1) space complexity
    fun detectCapitalUse(word: String): Boolean {
        if (word.length == 1) return true

        if (word[0].isUpperCase() && word[1].isUpperCase()) {
            for (i in 2 until word.length) {
                if (word[i].isLowerCase()) {
                    return false
                }
            }
        } else {
            for (i in 1 until word.length) {
                if (word[i].isUpperCase()) {
                    return false
                }
            }
        }
        return true
    }
}